plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -563,6 +658,81 @@ private static ConnectorPageSource createParquetPageSource(
         }
     }
 
+    private static class IcebergProjectedLayout
+            implements ProjectedLayout
+    {
+        private final Map<Integer, ProjectedLayout> projectedLayoutForFieldId;
+
+        private IcebergProjectedLayout(Map<Integer, ProjectedLayout> projectedLayoutForFieldId)
+        {
+            this.projectedLayoutForFieldId = projectedLayoutForFieldId;
+        }
+
+        public static ProjectedLayout createProjectedLayout(OrcColumn root, List<List<Integer>> fieldIdDereferences)
+        {
+            if (fieldIdDereferences.stream().map(List::size).anyMatch(java.util.function.Predicate.isEqual(0))) {
+                return fullyProjectedLayout();
+            }
+
+            Map<Integer, List<List<Integer>>> dereferencesByField = fieldIdDereferences.stream().collect(
+                    Collectors.groupingBy(
+                            sequence -> sequence.get(0),
+                            mapping(sequence -> sequence.subList(1, sequence.size()), toList())));
```

Comment:
It seems we're doing tree traversal here (effectively). 
wonder whether we can improve representation of dereferences so that we don't need to copy all the elements when recursing (OK to address in follow-up)